File: swift/LoopflowCore/Models/Step.swift
// Step model - the basic unit of execution.
// A Step is a prompt to run with optional config.

import Foundation

/// Per-step configuration overrides.
public struct StepConfig: Sendable, Codable, Equatable {
    public var model: String?
    public var defaultModel: String?
    public var direction: String?
    public var context: [String]?

    public init(
        model: String? = nil,
        defaultModel: String? = nil,
        direction: String? = nil,
        context: [String]? = nil
    ) {
        self.model = model
        self.defaultModel = defaultModel
        self.direction = direction
        self.context = context
    }

    public var isEmpty: Bool {
        model == nil &&
            defaultModel == nil &&
            direction == nil &&
            (context == nil || context!.isEmpty)
    }

    enum CodingKeys: String, CodingKey {
        case model
        case defaultModel = "default_model"
        case direction
        case context
    }
}

/// A Step is the basic unit - a prompt to run with optional config.
public struct Step: Sendable, Codable, Equatable, Identifiable {
    public var id: UUID = UUID()
    public var prompt: String
    public var config: StepConfig?

    enum CodingKeys: String, CodingKey {
        case prompt, config
    }

    public init(prompt: String, config: StepConfig? = nil) {
        self.prompt = prompt
        self.config = config
    }

    public init(from decoder: Decoder) throws {
        // Handle string shorthand: "design" -> Step(prompt: "design")
        if let singleValue = try? decoder.singleValueContainer(),
           let promptName = try? singleValue.decode(String.self) {
            self.prompt = promptName
            return
        }

        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.prompt = try container.decode(String.self, forKey: .prompt)
        self.config = try container.decodeIfPresent(StepConfig.self, forKey: .config)
    }

}

// MARK: - Step Execution

/// StepRun represents a single execution of a step.
/// Schema matches Python's lfd/models.py StepRun class.
public struct StepRun: Sendable, Identifiable, Codable, Hashable {
    public let id: String
    public let step: String  // matches Python schema (was "prompt")
    public let repo: String
    public let worktree: String
    public let status: String
    public let startedAt: Date
    public let endedAt: Date?
    public let model: String
    public let runMode: String

    enum CodingKeys: String, CodingKey {
        case id, step, repo, worktree, status, model
        case startedAt = "started_at"
        case endedAt = "ended_at"
        case runMode = "run_mode"
    }

    public var isRunning: Bool {
        status == "running" || status == "waiting"
    }

    public var isCompleted: Bool {
        status == "completed"
    }

    public var isError: Bool {
        status == "error"
    }

    /// Memberwise initializer (required since we add a custom init).
    public init(
        id: String,
        step: String,
        repo: String,
        worktree: String,
        status: String,
        startedAt: Date,
        endedAt: Date?,
        model: String,
        runMode: String
    ) {
        self.id = id
        self.step = step
        self.repo = repo
        self.worktree = worktree
        self.status = status
        self.startedAt = startedAt
        self.endedAt = endedAt
        self.model = model
        self.runMode = runMode
    }
}

File: web/src/models/step.ts
// Step model. Maps to swift/LoopflowCore/Models/Step.swift
// A Step is the basic unit - a prompt to run with optional config.

export interface StepConfig {
  model?: string
  voice?: string
  context?: string[]
}

export interface Step {
  id: string
  prompt: string
  config?: StepConfig
}

// Helper functions
export function stepConfigIsEmpty(config?: StepConfig): boolean {
  if (!config) return true
  return (
    config.model === undefined &&
    config.voice === undefined &&
    (config.context === undefined || config.context.length === 0)
  )
}

export function stepHasConfig(step: Step): boolean {
  return !stepConfigIsEmpty(step.config)
}

// StepRun = a single execution of a step

export type StepStatus = 'running' | 'completed' | 'error'

export interface StepRun {
  id: string
  step: string  // matches Python schema
  repo: string
  worktree: string  // matches Python schema (was worktreePath)
  status: StepStatus
  startedAt: Date
  endedAt?: Date
  model: string
  runMode: string
}

export function stepRunIsCompleted(run: StepRun): boolean {
  return run.status === 'completed'
}

export function stepRunIsRunning(run: StepRun): boolean {
  return run.status === 'running'
}

export interface FileChange {
  id: string
  path: string
  kind: 'added' | 'modified' | 'deleted' | 'renamed'
  linesAdded: number
  linesRemoved: number
  diffPreview?: string
}

export interface StepResult {
  id: string
  step: string  // matches Python schema
  worktree: string
  status: StepStatus
  startedAt: Date
  endedAt: Date
  filesChanged: FileChange[]
  newCommits: string[]
  baselineSHA: string
  currentSHA: string
}

// Computed properties
export function duration(result: StepResult): number {
  return (result.endedAt.getTime() - result.startedAt.getTime()) / 1000
}

export function hasChanges(result: StepResult): boolean {
  return result.filesChanged.length > 0 || result.newCommits.length > 0
}

export function totalLinesAdded(result: StepResult): number {
  return result.filesChanged.reduce((sum, f) => sum + f.linesAdded, 0)
}

export function totalLinesRemoved(result: StepResult): number {
  return result.filesChanged.reduce((sum, f) => sum + f.linesRemoved, 0)
}
